File: chronicle-pc/pipeline/chronicle_pipeline/api/kb.py
"""Knowledge-base notes markdown CRUD + tree (PARA-only after dual-read cutover)."""

from __future__ import annotations

import logging
from pathlib import Path
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from .. import path_map
from ..cutover_kb import MIGRATE_HINT
from ..link_repair import repair_links_after_move
from ..lock import vault_process_lock
from ..paths import atomic_write_text, content_hash
from .deps import get_root

log = logging.getLogger("chronicle.api.kb")

router = APIRouter(tags=["kb"])


class NoteBody(BaseModel):
    content: str = ""
    title: str | None = None
    tags: list[str] = Field(default_factory=list)
    section: str | None = None
    """Optional 'kb' | 'notes' — when creating a bare filename, picks the default area.
    When set, cross-section explicit PARA paths are rejected (400)."""
    base_hash: str | None = Field(
        default=None,
        pattern=r"^[0-9a-f]{64}$",
        description="Required on overwrite PUT; must match on-disk content_hash",
    )


class MoveBody(BaseModel):
    from_path: str
    to_path: str


class ArchiveBody(BaseModel):
    path: str


def _reject_legacy_or_400(path: str) -> str:
    """Normalize path; 410 for retired kb/notes; 400 for other invalid."""
    try:
        normalized = path_map.normalize_api_path(path)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e
    if path_map.is_legacy_kb_path(normalized):
        raise HTTPException(410, MIGRATE_HINT)
    try:
        return path_map.validate_knowledge_rel(normalized)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e


def _norm_or_400(path: str) -> str:
    return _reject_legacy_or_400(path)


def _section_or_400(section: str | None) -> str | None:
    try:
        return path_map.validate_section(section)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e


def _assert_section_or_400(rel: str, section: str | None) -> None:
    try:
        path_map.assert_path_allowed_for_section(rel, section)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e


def _is_bare_create_alias(raw_path: str) -> bool:
    """True when the client sent a bare relative path (not PARA, not legacy)."""
    p = raw_path.strip().lstrip("/").replace("\\", "/")
    while "//" in p:
        p = p.replace("//", "/")
    if path_map.is_para_prefix(p) or path_map.is_legacy_kb_path(p):
        return False
    if p.startswith("ResumePoints/"):
        return False
    return bool(p)


def _resolve_write_with_section(
    root: Path,
    rel: str,
    section: str | None,
    *,
    create: bool,
    bare_alias: bool = False,
) -> Path:
    """Like path_map.resolve_write, but honors an explicit create-time section
    for bare filenames remapped to Inbox (not ResumePoints).

    Section remap applies **only on create** for bare aliases. Explicit PARA
    paths are section-checked immediately (cross-section → 400).
    """
    section = path_map.validate_section(section)
    # Bare aliases normalize to Inbox before section remap — do not section-check
    # the intermediate Inbox path. Explicit PARA paths are checked immediately.
    if section and path_map.is_para_prefix(rel) and not bare_alias:
        _assert_section_or_400(rel, section)
    if not create:
        return path_map.resolve_write(root, rel, create=False)
    if (
        create
        and bare_alias
        and section
        and rel.startswith("00-Inbox/")
        and "ResumePoints" not in rel
    ):
        suffix = rel[len("00-Inbox/") :]
        dest_rel = f"{path_map.default_create_area(section)}/{suffix}"
        return path_map.abs_under_root(root, dest_rel)
    return path_map.resolve_write(root, rel, create=True)


def _title_for_create(rel: str, body: NoteBody) -> str | None:
    """Prefer explicit body.title; fall back to filename stem (not a path)."""
    if body.title is not None and body.title.strip():
        t = body.title.strip()
        # SPA historically passed the vault path as title — use stem instead.
        if "/" in t or t.endswith(".md"):
            return Path(t).stem
        return t
    return Path(rel).stem


def _content_for_create(rel: str, body: NoteBody) -> str:
    """Apply optional title/tags, then convention-complete create frontmatter."""
    from ..note_frontmatter import ensure_create_frontmatter

    content = body.content
    if not content.lstrip().startswith("---") and (body.title is not None or body.tags):
        lines = ["---"]
        if body.title is not None:
            lines.append(f"title: {_title_for_create(rel, body)}")
        if body.tags:
            tag_list = ", ".join(body.tags)
            lines.append(f"tags: [{tag_list}]")
        lines.append("---")
        lines.append("")
        content = "\n".join(lines) + content
    return ensure_create_frontmatter(
        content,
        title=_title_for_create(rel, body),
        note_type="note",
    )


def _maybe_frontmatter(body: NoteBody) -> str:
    """Legacy helper for overwrite paths that only inject title/tags when bare."""
    content = body.content
    if body.title is None and not body.tags:
        return content
    if content.lstrip().startswith("---"):
        return content
    lines = ["---"]
    if body.title is not None:
        lines.append(f"title: {body.title}")
    if body.tags:
        tag_list = ", ".join(body.tags)
        lines.append(f"tags: [{tag_list}]")
    lines.append("---")
    lines.append("")
    return "\n".join(lines) + content


def _para_suffix(rel: str) -> str:
    """Area-relative suffix for a knowledge path (for archive nesting)."""
    for area in path_map.PARA_AREAS:
        if rel == area or rel.startswith(area + "/"):
            return rel[len(area) + 1 :] if rel.startswith(area + "/") else Path(rel).name
    return Path(rel).name


def _move_note(root: Path, from_rel: str, to_rel: str) -> dict[str, Any]:
    """Move a PARA note to [to_rel]."""
    if not path_map.is_para_prefix(to_rel):
        raise HTTPException(400, "to_path must be under a PARA area")
    if path_map.is_chrome_path(to_rel):
        raise HTTPException(400, "cannot move onto chrome basename")

    abs_src = path_map.resolve_read_abs(root, from_rel)
    if abs_src is None or not abs_src.is_file():
        raise HTTPException(404, f"note not found: {from_rel}")
    primary_rel = path_map.vault_rel(root, abs_src)

    dest = path_map.abs_under_root(root, to_rel)
    if dest.exists() and dest.resolve() != abs_src.resolve():
        raise HTTPException(409, f"destination exists: {to_rel}")

    content = abs_src.read_text(encoding="utf-8", errors="replace")
    if dest.resolve() != abs_src.resolve():
        atomic_write_text(dest, content if content.endswith("\n") else content + "\n")
        abs_src.unlink()

    repair = repair_links_after_move(root, primary_rel, to_rel)
    return {
        "ok": True,
        "from_path": primary_rel,
        "to_path": to_rel,
        "quarantined": [],
        "links_repaired": repair.replacements,
        "files_updated": repair.files_updated,
        "changelog_appended": repair.changelog_appended,
    }


def _archive_dest_rel(rel: str) -> str:
    """90-Archive/<original subpath> — strip leading 90-Archive/ if already archived."""
    suffix = _para_suffix(rel)
    if path_map.is_para_prefix(rel) and rel.startswith("90-Archive/"):
        inner = rel[len("90-Archive/") :]
        if inner.startswith("_legacy-kb/"):
            return rel
        return f"90-Archive/{inner}"
    return f"90-Archive/{suffix}"


HUB_READ_PATHS: frozenset[str] = frozenset({"Home.md"})


def _norm_read_path(path: str) -> str:
    """Validate path for GET; allows vault-root hub files outside PARA."""
    p = path.strip().lstrip("/")
    if p in HUB_READ_PATHS:
        return p
    return _norm_or_400(p)


@router.get("/kb/templates")
def kb_templates(root: Path = Depends(get_root)) -> dict[str, Any]:
    """List ``_templates/*.md`` for the SPA create picker."""
    templates_dir = root / "_templates"
    files: list[dict[str, str]] = []
    if templates_dir.is_dir():
        for path in sorted(templates_dir.glob("*.md")):
            content = path.read_text(encoding="utf-8", errors="replace")
            files.append(
                {
                    "name": path.stem,
                    "path": f"_templates/{path.name}",
                    "content": content,
                }
            )
    return {"files": files}


@router.get("/kb/tree")
def kb_tree(
    root: Path = Depends(get_root),
    section: str | None = Query(default=None, description="'kb' or 'notes'; omit for all"),
) -> dict[str, Any]:
    section = _section_or_400(section)
    try:
        tree = path_map.build_knowledge_tree(root, section=section)
    except ValueError as e:
        raise HTTPException(400, str(e)) from e
    files = path_map.list_knowledge_files(root)
    if section:
        files = [f for f in files if path_map.section_for(f) == section]
    return {"tree": tree, "files": files}


def _note_response(abs_path: Path, root: Path, *, path_override: str | None = None) -> dict[str, Any]:
    content = abs_path.read_text(encoding="utf-8", errors="replace")
    rel = path_override or path_map.vault_rel(root, abs_path)
    return {
        "path": rel,
        "content": content,
        "content_hash": content_hash(content),
    }


@router.get("/kb/notes/{path:path}")
def get_kb_note(path: str, root: Path = Depends(get_root)) -> dict[str, Any]:
    rel = _norm_read_path(path)
    if rel in HUB_READ_PATHS:
        abs_path = root / rel
        if not abs_path.is_file():
            raise HTTPException(404, f"note not found: {rel}")
        return _note_response(abs_path, root, path_override=rel)
    abs_path = path_map.resolve_read_abs(root, rel)
    if abs_path is None or not abs_path.is_file():
        raise HTTPException(404, f"note not found: {rel}")
    return _note_response(abs_path, root)


@router.put("/kb/notes/{path:path}")
def put_kb_note(
    path: str, body: NoteBody, root: Path = Depends(get_root)
) -> dict[str, Any]:
    """Create or overwrite a knowledge note (PARA only).

    Overwrite requires ``base_hash`` matching on-disk ``content_hash`` (409 on mismatch).
    Creates ignore ``base_hash``.
    """
    with vault_process_lock(root):
        bare = _is_bare_create_alias(path)
        rel = _norm_or_400(path)
        section = _section_or_400(body.section)
        existing = path_map.resolve_read_abs(root, rel)
        creating = existing is None or not existing.is_file()
        if creating and section:
            abs_probe = _resolve_write_with_section(
                root, rel, section, create=True, bare_alias=bare
            )
            dest_rel = path_map.vault_rel(root, abs_probe)
            _assert_section_or_400(dest_rel, section)
        abs_path = _resolve_write_with_section(
            root, rel, section, create=False, bare_alias=bare
        )
        if not abs_path.is_file():
            abs_path = _resolve_write_with_section(
                root, rel, section, create=True, bare_alias=bare
            )
            if section:
                _assert_section_or_400(path_map.vault_rel(root, abs_path), section)
            creating = True
        if not creating:
            disk = abs_path.read_text(encoding="utf-8", errors="replace")
            disk_hash = content_hash(disk)
            if not body.base_hash:
                raise HTTPException(
                    400,
                    "base_hash required for overwrite (from GET content_hash)",
                )
            if disk_hash != body.base_hash:
                raise HTTPException(
                    409,
                    {
                        "detail": "knowledge note hash mismatch",
                        "on_disk_hash": disk_hash,
                    },
                )
        text = (
            _content_for_create(path_map.vault_rel(root, abs_path), body)
            if creating
            else _maybe_frontmatter(body)
        )
        atomic_write_text(abs_path, text if text.endswith("\n") else text + "\n")
        return _note_response(abs_path, root)


@router.post("/kb/notes/{path:path}", status_code=201)
def post_kb_note(
    path: str, body: NoteBody, root: Path = Depends(get_root)
) -> dict[str, Any]:
    """Create a knowledge note; fails if it already exists."""
    with vault_process_lock(root):
        bare = _is_bare_create_alias(path)
        rel = _norm_or_400(path)
        section = _section_or_400(body.section)
        existing = path_map.resolve_read_abs(root, rel)
        if existing is not None and existing.is_file():
            raise HTTPException(
                409, f"note already exists: {path_map.vault_rel(root, existing)}"
            )
        abs_path = _resolve_write_with_section(
            root, rel, section, create=True, bare_alias=bare
        )
        if section:
            _assert_section_or_400(path_map.vault_rel(root, abs_path), section)
        if abs_path.is_file():
            raise HTTPException(
                409, f"note already exists: {path_map.vault_rel(root, abs_path)}"
            )
        text = _content_for_create(path_map.vault_rel(root, abs_path), body)
        atomic_write_text(abs_path, text if text.endswith("\n") else text + "\n")
        return _note_response(abs_path, root)


@router.delete("/kb/notes/{path:path}")
def delete_kb_note(path: str, root: Path = Depends(get_root)) -> dict[str, Any]:
    """Hard-delete a PARA note. Prefer archive."""
    with vault_process_lock(root):
        rel = _norm_or_400(path)
        abs_path = path_map.resolve_read_abs(root, rel)
        if abs_path is None or not abs_path.is_file():
            raise HTTPException(404, f"note not found: {rel}")
        deleted_rel = path_map.vault_rel(root, abs_path)
        abs_path.unlink()
        return {"ok": True, "deleted": deleted_rel, "deleted_all": [deleted_rel]}


@router.post("/kb/move")
def move_kb_note(body: MoveBody, root: Path = Depends(get_root)) -> dict[str, Any]:
    """Move a note to a PARA path."""
    with vault_process_lock(root):
        from_rel = _norm_or_400(body.from_path)
        to_rel = _norm_or_400(body.to_path)
        return _move_note(root, from_rel, to_rel)


@router.post("/kb/archive")
def archive_kb_note(body: ArchiveBody, root: Path = Depends(get_root)) -> dict[str, Any]:
    """Archive under 90-Archive/<original subpath>/ (preferred over hard delete)."""
    with vault_process_lock(root):
        from_rel = _norm_or_400(body.path)
        to_rel = _archive_dest_rel(from_rel)
        to_rel = _norm_or_400(to_rel)
        return _move_note(root, from_rel, to_rel)

File: chronicle-pc/server/src/kb.rs
//! PARA knowledge CRUD + move/archive (api/kb.py logic as library functions).

use std::path::Path;

use serde_json::{json, Value};

use crate::errors::{ApiError, ChronicleError};
use crate::frontmatter;
use crate::link_repair;
use crate::lock::vault_lock;
use crate::models::Entry;
use crate::path_map;
use crate::paths::content_hash;

pub fn read_note_abs(root: &Path, rel: &str) -> Result<std::path::PathBuf, ApiError> {
    // Legacy 410.
    if path_map::is_legacy_kb_path(&path_map::_norm(rel)) {
        return Err(ApiError::new(
            axum::http::StatusCode::GONE,
            json!(path_map::MIGRATE_HINT),
        ));
    }
    match path_map::resolve_read_abs(root, rel) {
        Ok(Some(p)) => Ok(p),
        Ok(None) => Err(ApiError::not_found(format!("note not found: {rel}"))),
        Err(e) => Err(e.into()),
    }
}

pub fn get_note(root: &Path, rel_raw: &str) -> Result<Value, ApiError> {
    // Home.md hub exception at vault root.
    if path_map::_norm(rel_raw) == "Home.md" {
        let p = root.join("Home.md");
        if p.is_file() {
            let text = std::fs::read_to_string(&p).unwrap_or_default();
            return Ok(json!({
                "path": "Home.md",
                "content": text,
                "content_hash": content_hash(&text),
            }));
        }
    }
    let rel = validate_rel(rel_raw)?;
    let abs = read_note_abs(root, &rel)?;
    let bytes = std::fs::read(&abs).map_err(|_| ApiError::not_found(format!("note not found: {rel}")))?;
    let text = String::from_utf8_lossy(&bytes).to_string();
    Ok(json!({
        "path": rel,
        "content": text,
        "content_hash": content_hash(&text),
    }))
}

fn validate_rel(rel_raw: &str) -> Result<String, ApiError> {
    path_map::validate_knowledge_rel(rel_raw).map_err(|e| ApiError::bad_request(e.to_string()))
}

fn inject_frontmatter(body: &str, title: Option<&str>) -> String {
    let mut out = body.to_string();
    if !out.trim_start().starts_with("---") {
        let mut lines: Vec<String> = Vec::new();
        if let Some(t) = title.map(str::trim).filter(|t| !t.is_empty()) {
            lines.push(format!("title: {t}"));
        }
        if !lines.is_empty() {
            out = format!("---\n{}\n---\n{}", lines.join("\n"), out);
        }
    }
    frontmatter::ensure_create_frontmatter(&out, title, "note")
}

pub struct WriteNoteArgs<'a> {
    pub rel_raw: &'a str,
    pub content: &'a str,
    pub base_hash: Option<&'a str>,
    pub section: Option<&'a str>,
    pub create: bool,
}

pub fn write_note(root: &Path, args: WriteNoteArgs<'_>) -> Result<Value, ApiError> {
    let _guard = crate::lock::vault_lock(root, Some(std::time::Duration::from_secs(30))).map_err(ApiError::from)?;

    let normalized = path_map::normalize_api_path(args.rel_raw).map_err(|e| ApiError::bad_request(e.to_string()))?;
    let is_bare_alias = !path_map::is_para_prefix(&normalized)
        && !path_map::is_legacy_kb_path(&normalized)
        && !normalized.starts_with("ResumePoints/")
        && !normalized.is_empty();

    let mut rel = validate_rel(args.rel_raw)?;
    if args.create && is_bare_alias {
        if let Some(section) = args.section {
            let area = path_map::default_create_area(section);
            let suffix = normalized.trim_start_matches('/');
            rel = format!("{area}/{suffix}");
        }
    }
    if let Some(section) = args.section {
        let sec = path_map::validate_section(Some(section)).map_err(|e| ApiError::bad_request(e.to_string()))?;
        path_map::assert_path_allowed_for_section(&rel, sec.as_deref()).map_err(|e| ApiError::bad_request(e.to_string()))?;
    }

    let abs = path_map::resolve_write(root, &rel).map_err(|e| ApiError::bad_request(e.to_string()))?;
    let existed = abs.is_file();

    if existed && !args.create {
        let disk_text = std::fs::read_to_string(&abs).unwrap_or_default();
        let Some(base) = args.base_hash else {
            return Err(ApiError::bad_request("base_hash required for overwrite (from GET content_hash)"));
        };
        if content_hash(&disk_text) != base {
            return Err(ApiError::conflict_object(json!({
                "detail": "knowledge note hash mismatch",
                "on_disk_hash": content_hash(&disk_text),
            })));
        }
        let mut text = args.content.to_string();
        if !text.ends_with('\n') {
            text.push('\n');
        }
        crate::paths::atomic_write_text(&abs, &text)?;
        return Ok(json!({
            "path": rel,
            "content": text,
            "content_hash": content_hash(&text),
        }));
    }
    if args.create && existed {
        return Err(ApiError::conflict(format!("note already exists: {rel}")));
    }

    // Create pipeline.
    let title = Path::new(&rel)
        .file_stem()
        .map(|s| s.to_string_lossy().to_string());
    let full = inject_frontmatter(args.content, title.as_deref());
    let mut text = full;
    if !text.ends_with('\n') {
        text.push('\n');
    }
    std::fs::create_dir_all(abs.parent().unwrap()).map_err(|e| ApiError::internal(e.to_string()))?;
    crate::paths::atomic_write_text(&abs, &text)?;
    Ok(json!({
        "path": rel,
        "content": text,
        "content_hash": content_hash(&text),
    }))
}

pub fn delete_note(root: &Path, rel_raw: &str) -> Result<Value, ApiError> {
    let _guard = crate::lock::vault_lock(root, Some(std::time::Duration::from_secs(30))).map_err(ApiError::from)?;
    let rel = validate_rel(rel_raw)?;
    let abs = read_note_abs(root, &rel)?;
    std::fs::remove_file(&abs).map_err(|e| ChronicleError::Io(e.to_string()))?;
    Ok(json!({"ok": true, "deleted": rel, "deleted_all": [rel]}))
}

fn archive_dest_rel(rel: &str) -> String {
    if let Some(inner) = rel.strip_prefix("90-Archive/") {
        if inner.starts_with("_legacy-kb/") {
            return format!("90-Archive/{inner}");
        }
    }
    if rel == "90-Archive" || rel.starts_with("90-Archive/") {
        return rel.to_string();
    }
    for area in path_map::PARA_AREAS {
        if let Some(suffix) = rel.strip_prefix(&format!("{area}/")) {
            return format!("90-Archive/{suffix}");
        }
        if rel == area {
            return "90-Archive".to_string();
        }
    }
    format!("90-Archive/{}", rel)
}

/// move/archive share the machinery; `archive=true` derives to_path.
pub fn move_note(
    root: &Path,
    from_raw: &str,
    to_raw: Option<&str>,
    archive: bool,
) -> Result<Value, ApiError> {
    let _guard = crate::lock::vault_lock(root, Some(std::time::Duration::from_secs(30))).map_err(ApiError::from)?;
    let from_rel = validate_rel(from_raw)?;
    let to_rel = if archive {
        let d = archive_dest_rel(&from_rel);
        if !d.ends_with(".md") {
            return Err(ApiError::bad_request("cannot archive a non-note"));
        }
        path_map::validate_knowledge_rel(&d).map_err(|e| ApiError::bad_request(e.to_string()))?
    } else {
        let raw = to_raw.ok_or_else(|| ApiError::bad_request("to_path is required"))?;
        let t = validate_rel(raw)?;
        if !path_map::is_para_prefix(&t) {
            return Err(ApiError::bad_request("to_path must be under a PARA area"));
        }
        let basename = Path::new(&t)
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default();
        if path_map::is_chrome_basename(&basename) {
            return Err(ApiError::bad_request("cannot move onto chrome basename"));
        }
        t
    };

    let src = read_note_abs(root, &from_rel)?;
    if to_rel == from_rel {
        // no-op move
    } else {
        let dst = root.join(&to_rel);
        if dst.is_file() {
            return Err(ApiError::conflict(format!("destination exists: {to_rel}")));
        }
    }

    let bytes = std::fs::read(&src).map_err(|_| ApiError::not_found(format!("note not found: {from_rel}")))?;
    let mut text = String::from_utf8_lossy(&bytes).to_string();
    if !text.ends_with('\n') {
        text.push('\n');
    }
    let dst = root.join(&to_rel);
    if let Some(parent) = dst.parent() {
        std::fs::create_dir_all(parent).map_err(|e| ApiError::internal(e.to_string()))?;
    }
    crate::paths::atomic_write_text(&dst, &text)?;
    std::fs::remove_file(&src).map_err(|e| ChronicleError::Io(e.to_string()))?;

    let (links, files) =
        link_repair::repair_links(root, &from_rel, &to_rel).map_err(ApiError::from)?;
    link_repair::append_changelog(root, &from_rel, &to_rel, links, files).map_err(ApiError::from)?;

    Ok(json!({
        "ok": true,
        "from_path": from_rel,
        "to_path": to_rel,
        "quarantined": [],
        "links_repaired": links,
        "files_updated": files,
        "changelog_appended": true,
    }))
}

pub fn templates(root: &Path) -> Value {
    let dir = root.join("_templates");
    let mut files: Vec<std::path::PathBuf> = std::fs::read_dir(&dir)
        .map(|rd| {
            rd.flatten()
                .map(|e| e.path())
                .filter(|p| p.extension().and_then(|x| x.to_str()) == Some("md"))
                .collect()
        })
        .unwrap_or_default();
    files.sort();
    let out: Vec<Value> = files
        .iter()
        .filter_map(|p| {
            let name = p.file_stem()?.to_string_lossy().to_string();
            let rel_name = p.file_name()?.to_string_lossy().to_string();
            let bytes = std::fs::read(p).ok()?;
            Some(json!({
                "name": name,
                "path": format!("_templates/{rel_name}"),
                "content": String::from_utf8_lossy(&bytes),
            }))
        })
        .collect();
    json!({ "files": out })
}

/// Entry helper reused by api::entries — parse ts or generate now.
pub(crate) fn entry_from_create(
    root: &Path,
    id_in: Option<String>,
    ts_in: Option<String>,
    kind: &str,
    text: &str,
    tags: Vec<String>,
    mood: Option<i64>,
    images: Vec<String>,
    audio: Vec<String>,
) -> Result<Entry, ApiError> {
    use chrono::{DateTime, FixedOffset, Local};
    let _ = Local;
    let ts = match ts_in.as_deref().filter(|s| !s.trim().is_empty()) {
        Some(s) => match crate::timeutil::parse_iso_aware(s) {
            Some((dt, false)) => dt.to_rfc3339_opts(chrono::SecondsFormat::Secs, false),
            Some((_dt, true)) => s.trim().to_string(),
            None => return Err(ApiError::bad_request(format!("invalid ts: {s:?}"))),
        },
        None => {
            let now: DateTime<FixedOffset> = Local::now().into();
            now.to_rfc3339_opts(chrono::SecondsFormat::Secs, false)
        }
    };
    let id = match id_in {
        Some(id) => {
            crate::entries::validate_id(&id).map_err(ApiError::from)?;
            if !id.ends_with("-pc") && !id.contains("-pc_") {
                return Err(ApiError::bad_request("API writes must use -pc entry ids"));
            }
            if crate::entries::entry_path(root, &id).is_ok_and(|p| p.is_file()) {
                return Err(ApiError::conflict(format!("entry already exists: {id}")));
            }
            id
        }
        None => crate::entries::next_pc_id(root, chrono::Local::now().fixed_offset()),
    };
    if let Some(m) = mood {
        if !(1..=5).contains(&m) {
            return Err(ApiError::bad_request("mood must be 1–5 or null"));
        }
    }
    Ok(Entry {
        version: 1,
        id,
        ts,
        kind: kind.to_string(),
        text: text.to_string(),
        tags,
        images,
        audio,
        mood,
        processed: false,
        filed: false,
        filed_content_hash: None,
        filed_path: None,
        prose_edited: false,
        extra: Default::default(),
    })
}
